### PR TITLE
Issue 18407 - debug should escape nothrow

### PR DESCRIPTION
**Pull request by wilzbach**
References:
- https://issues.dlang.org/show_bug.cgi?id=18407
- https://github.com/dlang/dmd/pull/7882
- https://github.com/dlang/dlang.org/pull/2209

I wasn't entirely sure on the best way to do this.
The problem is that `blockexit` checker runs so late in semantic3 that the DebugStatements are already flattened out.

I'm not proud of the current solution, so I'm more than open to better suggestions.

### CODE REVIEW

File: changelog/debug-nothrow.dd [hunk 1]
```diff
@@ -0,0 +1,11 @@
+Unsafe code can now be used in nothrow blocks
```
Comment (jacob-carlborg) [inline changelog/debug-nothrow.dd]: This isn’t related to safe code?

File: src/dmd/statement.d [hunk 2]
```diff
@@ -1448,6 +1451,45 @@ extern (C++) final class ConditionalStatement : Statement
     }
 }
 
+/**
+Marks all occurring ThrowStatements as as internalThrows.
```
Comment (RazvanN7) [inline src/dmd/statement.d]: double **as**

**Review by RazvanN7 — APPROVED** [review]
LGTM

File: src/dmd/statement.d [hunk 3]
```diff
@@ -1448,6 +1451,45 @@ extern (C++) final class ConditionalStatement : Statement
...
+    public:
+
+        override void visit(ThrowStatement s)
+        {
+            s.internalThrow = true;
+        }
+
+        override void visit(TryCatchStatement s)
+        {
+            foreach (c; *s.catches)
```
Comment (RazvanN7) [inline src/dmd/statement.d]: uncovered

File: src/dmd/statement.d [hunk 4]
```diff
@@ -1448,6 +1451,45 @@ extern (C++) final class ConditionalStatement : Statement
...
+            {
+                c.internalCatch = true;
```
Comment (RazvanN7) [inline src/dmd/statement.d]: uncovered

**Comment by RazvanN7** [comment]
Actually, I just thought about something: what about anonymous functions that get called in debug statements?

```d
void foo() nothrow
{
    debug
    {
        () (throw new Exception("");)()
    }
}
```

Does this work with your patch? If, yes please add an example.

File: changelog/debug-nothrow.dd [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+When writing debug code, one isn't interested in the type safety, but a pleasant
+debugging experience.
+The type checker already allowed to escape `pure` and `@nogc` within `debug`
+statement.
+With this release, `nothrow` code can be called from `debug` statements too:
+
+---
+void main()
+{
+    debug throw new Exception("");
```
Comment (joakim-noah) [inline changelog/debug-nothrow.dd]: Nothing exceptional about this, did you mean to make the main `nothrow`?

File: changelog/debug-nothrow.dd [hunk 6]
```diff
@@ -0,0 +1,14 @@
+Nothrow code can now be used in debug blocks
+
+When writing debug code, one isn't interested in the type safety, but a pleasant
```
Comment (jacob-carlborg) [inline changelog/debug-nothrow.dd]: I don’t think `nothrow` has any thing to do with type safety. This is clearly a copy paste from the other PR for `@safe`.

**Review by jacob-carlborg — DISMISSED** [review]

**Comment by wilzbach** [comment]
BTW the following will be invalid after this PR.
I could add `inDebugStatement` to `NewExp` et.al. too, but I already feel like adding it to `CallExp` is wrong.

```d
void test6() nothrow
{
    static struct S
    {
        this(this)
        {
            throw new Exception("BOOM!");
        }
    }

    S    ss;

    // TOKassign
    ss = ss;

    // TOKconstruct
    S    ss2 = ss;
}

void test7() nothrow
{
    static struct S
    {
        this(int k)
        {
            throw new Exception("BOOM!");
        }
    }
    static class C
    {
        this(int k)
        {
            throw new Exception("BOOM!");
        }
    }
    debug S(10);
    debug new C(10);
}
```

File: src/dmd/mars.d [hunk 7]
```diff
@@ -498,6 +498,14 @@ private int tryMain(size_t argc, const(char)** argv)
     Objc._init();
     builtin_init();
 
+    version(CRuntime_Microsoft)
+    {
+        import dmd.root.longdouble;
+        initFPU();
+    }
+    import dmd.root.ctfloat : CTFloat;
+    CTFloat.initialize();
```
Comment (Geod24) [inline src/dmd/mars.d]: That doesn't look related ?

**Comment by WalterBright** [comment]
My concern with this is that statements get replaced with other statements in the semantic analysis, and those flags may get lost.

File: changelog/debug-nothrow.dd [hunk 8]
```diff
@@ -0,0 +1,14 @@
+Nothrow code can now be used in debug blocks
```
Comment (Geod24) [inline changelog/debug-nothrow.dd]: > Code in `debug` blocks will now ignore `nothrow`

Or something related. `nothrow` code can already be used in `debug` blocks :)

Comment (wilzbach) [reply changelog/debug-nothrow.dd]: Fixed :O

**Comment by wilzbach** [comment]
> My concern with this is that statements get replaced with other statements in the semantic analysis, and those flags may get lost.

Yes, good point, but I didn't know a better way to store this information as the `canthrow` checker runs late in the semantic3 analysis and thus the original `debug` statements have already been replaced.

**Comment by andralex** [comment]
status?

**Comment by wilzbach** [comment]
> status?

This is still ready, but @WalterBright wasn't fully happy with this.
As explained above, I don't know a better way to hack this into DMD as the information gets lost before we reach semantic3.

**Review by thewilsonator — APPROVED** [review]

**Review by jacob-carlborg — APPROVED** [review]

File: src/dmd/expression.d [hunk 9]
```diff
@@ -4696,6 +4696,7 @@ extern (C++) final class CallExp : UnaExp
     Expressions* arguments; // function arguments
     FuncDeclaration f;      // symbol to call
     bool directcall;        // true if a virtual call is devirtualized
+    bool inDebugStatement;  // true if this was in a debug statement
```
Comment (ibuclaw) [inline src/dmd/expression.d]: H...h...h...

**Review by ibuclaw — DISMISSED** [review]

File: src/dmd/expression.d (hunk 9 above)
Comment (WalterBright) [reply src/dmd/expression.d]: I presume you mean fix the corresponding .h file.

**Comment by quickfur** [comment]
ping @WalterBright 

It's been a *little* bit more than 72hrs now. ;-)  Any updates?

**Comment by 12345swordy** [comment]
It's green!

**Comment by Geod24** [comment]
Amazing

File: test/compilable/test16492.d [hunk 10]
```diff
@@ -40,3 +40,48 @@ void unsafeDIP1000Lifetime()(ref char[] p, scope char[] s)
...
+    debug {
+        foreach (_; 0 .. 10) {
+            if (1) {
+                throw new Exception("");
+            }
+        }
+    }
+}
+
+void test3() nothrow
```
Comment (ibuclaw) [inline test/compilable/test16492.d:61]: Broke the testsuite (function name conflict).
